### PR TITLE
feat(hermes): add additional sse features

### PR DESCRIPTION
**Pull request by cctdaniel**
- **add allow_unordered query param**
- **add benchmarks_only query param**
- **update docs**

### CODE REVIEW

File: hermes/src/api/sse.rs [hunk 1]
```diff
@@ -99,16 +108,30 @@ pub async fn price_stream_sse_handler(
...
+                        params.benchmarks_only,
                     )
                     .await
                     {
-                        Ok(price_update) => Ok(Event::default().json_data(price_update).unwrap()),
+                        Ok(price_update) => {
+                            // Check if there is any data to send
+                            if price_update.binary.data.is_empty() {
+                                // No data to send, skip creating an event
+                                return Ok(Event::default().comment("No data to send"));
```
Comment (ali-bahjati) [inline hermes/src/api/sse.rs:131]: What's the behaviour of this one? It sends an empty event? Or nothing?

File: hermes/src/api/sse.rs [hunk 2]
```diff
@@ -99,16 +108,30 @@ pub async fn price_stream_sse_handler(
...
+                            }
+                            Ok(Event::default().json_data(price_update).unwrap())
```
Comment (ali-bahjati) [inline hermes/src/api/sse.rs:133]: I'm a bit worried about having `unwraps`. can we add `unwrap_or_else(|e| error_event(e))`?

File: hermes/src/api/sse.rs [hunk 3]
```diff
@@ -126,18 +149,51 @@ async fn handle_aggregation_event(
     mut price_ids: Vec<PriceIdentifier>,
     encoding: EncodingType,
     parsed: bool,
+    benchmarks_only: bool,
```
Comment (ali-bahjati) [inline hermes/src/api/sse.rs:152]: nit: Is it a benchmark price or benchmarks price?

File: hermes/src/api/sse.rs [hunk 4]
```diff
@@ -99,16 +108,30 @@ pub async fn price_stream_sse_handler(
         async move {
             match message {
                 Ok(event) => {
+                    // Check if the event is out-of-order and if unordered updates are not allowed
+                    if let AggregationEvent::OutOfOrder { .. } = event {
+                        if !params.allow_unordered {
+                            return Ok(Event::default().comment("Out-of-order event skipped"));
+                        }
+                    }
```
Comment (ali-bahjati) [inline hermes/src/api/sse.rs:116]: I think it's cleaner if the handle_aggregation_event does this as well. We want the business logic to be all in one place. Then you can say that `handle_aggregation_event` returns `Result<Option<PriceUpdate>>` to also be more clear when it is going to skip the events.

File: hermes/src/api/sse.rs (hunk 1 above)
Comment (cctdaniel) [reply hermes/src/api/sse.rs:131]: this will send an SSE event with a comment, no data

File: hermes/src/api/sse.rs (hunk 4 above)
Comment (cctdaniel) [reply hermes/src/api/sse.rs:116]: agreed, will do

File: hermes/src/api/rest/v2/sse.rs [hunk 5]
```diff
@@ -105,10 +114,26 @@ pub async fn price_stream_sse_handler(
...
+                        params.benchmarks_only,
+                        params.allow_unordered,
                     )
                     .await
                     {
-                        Ok(price_update) => Ok(Event::default().json_data(price_update).unwrap()),
+                        Ok(price_update) => {
+                            // Check if there is any data to send
+                            if let Some(update) = price_update {
+                                if update.binary.data.is_empty() {
```
Comment (ali-bahjati) [inline hermes/src/api/rest/v2/sse.rs:125]: I think it might be better if we return a None if the list of price_ids becomes empty because of benchmarks.

**Pull request merged**